[DEV-37] husky를 사용하여 커밋과 푸시할 때 git hooks 설정

ono212 opened the pull request:
## 📌 기능 설명

- `husky`, `lint-staged` 라이브러리 설치
- `husky`로 `Git hooks` 설정

## 📌 구현 내용
<!-- 실제 구현 내용을 디테일하게 작성해 주세요 -->

<!-- ex)
- 피그마에 디자인된 사항으로 마크업을 구현했습니다.
- 헤더 컴포넌트에 필요한 이벤트를 hook으로 구현했습니다. 
-->

`pre-commit`훅과 `pre-push`훅을 설정해주었습니다.
- `pre-commit`훅은 커밋 명령어를 실행했을 때 커밋되기 전에 실행되는 훅으로, 커밋할 파일들을 대상으로 eslint와 prettier 규칙을 지키는지 확인합니다. 규칙에 맞지 않는 코드가 있을 경우 커밋되지 않습니다
- `pre-push`훅은 푸시 명령어를 실행했을 때 푸시되기 전에 실행되는 훅으로, `npm run build`를  실행하도록 하여 빌드 에러가 나지 않을 경우에만 푸시가 되도록 설정했습니다. 빌드 에러가 발생할 경우 푸시되지 않습니다!
---

husky 버전을 8.0.0에서 9.1.5버전으로 변경했습니다.
노트북 바꾼 후에 맥북에서 실행해보니까
```
hint: The 'pre-commit' hook was ignored because it's not set as executable.
hint: You can disable this warning with `git config advice.ignoredHook false`
```
위와 같은 오류로 hook이 실행되지 않아서 chmod를 해줘야한다고 해서 알아보던 중 [9버전 이상부터 chmod를 따로 설정해주지 않아도 hook이 정상 동작](https://github.com/typicode/husky/releases/tag/v9.0.1)하기때문에 버전업 해줬습니다!


## 📌 구현 결과
<!-- 구현 결과를 확인할 수 있는 방법 혹은 이미지를 첨부해주세요. -->

#### 1. lint오류가 있어서 커밋 실패할 경우
<img src="https://github.com/user-attachments/assets/0be410cd-0f0c-4d86-890d-d720d9ee7289" width="70%" height="90%" />

#### 2. lint검사를 통과해서 커밋 성공했을 경우
<img src="https://github.com/user-attachments/assets/d88a7318-3842-46bf-a752-c44a68ff3dc8" width="70%" height="90%" />

## 📌 논의하고 싶은 점
<!-- 논의하고 싶은 점이 있다면 적어주세요 -->
<!-- ex)
react-query 를 사용할 때 별도의 hook으로 만들어서 사용하시나요?
저는 별도로 사용하지 않는데 어떻게 하는게 좋을까요? 
-->

#### 1. 추가할 혹은 제거할 훅이 있을까요? `pre-push`훅만 있어도 될지 의견이 궁금합니다!!

#### 2. vscode의 소스제어로 git 사용하실 경우에 필요한 추가적인 설정
저는 주로 터미널보다는 vscode의 소스제어로 커밋을 하는데, vscode설정의 `setting.json`에 추가적인 설정을 해주지 않으면 훅이 아예 동작하지 않아서 문제있는 코드도 커밋 & 푸시하라구요..! (터미널로 git사용하시면 추가 설정안해도 정상 동작합니다!!!)

```json
{
  "git.enableSmartCommit": true, // 스마트 커밋 활성화 (옵션)
  "git.confirmSync": false // 동기화 확인 비활성화 (옵션)
}
```
![image](https://github.com/user-attachments/assets/57097d6c-b1dc-4e02-b59c-d8fe2103b73c)

이렇게 프로젝트 폴더 최상위의 `.vscode`폴더의 `setting.json`파일에 설정해주면, vscode의 소스제어로 git을 사용할 때도 훅이 정상 동작합니다!! 문제가 있을 경우 경고창으로 뜨는데 "git log열기"를 클릭하면 에러로그를 확인할 수 있습니다!
혹은 더 쉽게 설정하는 방법 있을까요?

ono212 committed:
feat: "잘못된 정보 문의하기" 버튼 UI 추가

ono212 committed:
feat: 검색 Input창에 "이 용어 제보하기" 버튼 UI 추가

ono212 committed:
feat: 검색결과가 없는 페이지에  "단어 제보하러 가기" 버튼 UI 추가

ono212 committed:
fix: 커밋 누락한 파일 추가

uk11 committed:
fix: css 수정

uk11 committed:
feat: > 아이콘 추가

uk11 committed:
fix: > 아이콘 color props 추가

uk11 committed:
fix: 단어 -> 용어 수정

uk11 committed:
feat: 용어 제보하기 url

uk11 committed:
fix: className 공백 제거

uk11 committed:
fix: 트렌딩 용어 -> 단어 수정

uk11 committed:
fix: 퀴즈 결과 페이지 헤더 문구 수정

ono212 committed:
refactor: 스타일 지정 수정 / gap-[8px] -> gap-2

ono212 committed:
chore: husky 설치 및 초기화

ono212 committed:
chore: lint-staged 라이브러리 설치,  hooks 설정

- 스테이징된 파일들을 커밋하기 전에 eslint, prettier 검사를 실행하도록 지정

ono212 committed:
feat: 피싱 공격 방지를 위해 참조를 없애는 rel속성 rel="noreferrer noopener" 추가

ono212 committed:
feat: 커밋 전에 eslint와 prettier검사를 하도록 설정

ono212 committed:
chore: .gitignore에 .eslintcache파일 추가

- eslint검사 시 캐시되도록 설정했기 때문에 캐시 파일이 생기고 있음 -> git에 포함시키지 않도록 추가

ono212 committed:
chore: 푸시하기 전에 빌드를 실행하도록 설정

ono212 committed:
Merge branch 'main' into chore/DEV-37

uk11 committed:
fix: 검색 결과가 없어요. 문구를 클릭해도 제보하기 폼으로 넘어가는 문제 수정

uk11 committed:
fix: 퀴즈 뒤로가기 모달 추가

ono212 committed:
chore: pwa파일 변경된 것

uk11 committed:
fix: 검색 페이지 메타데이터 url 수정

nyoung113 reviewed (commented):
👍 
와 드디어 허스키!!! 수고하셨어요 치치!!

nyoung113 commented on .husky/pre-push:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+# push하기 전에 빌드 실행
```
와 쵝오,, 이거 정말 필요했는데!! 감사합니다 !!! ☺️

ono212 replied:
감사합니다 돌맹😆! 제가 PR작성이 늦어서..! 늦게나마 작성했는데 혹시나 확인 한 번 부탁드려요..🥹

uk11 committed:
Update README.md

ono212 committed:
chore: husky라이브러리 버전 변경

- 9버전 이상인 최신 버전으로 변경
- https://github.com/typicode/husky/releases/tag/v9.0.1 9버전 이상부터 chmod를 따로 설정해주지 않기 때문에 버전업

ono212 committed:
chore: Husky의 이전 버전에서 필요했던 초기화 코드 제거

ono212 committed:
chore: pwa관련 파일 변경된 것

ono212 committed:
Merge branch 'main' into chore/DEV-37
